ext/opcache/Optimizer/dce.c

```diff
@@ -0,0 +1,647 @@
+static zend_bool dce_instr(context *ctx, zend_op *opline, zend_ssa_op *ssa_op) {
+	zend_ssa *ssa = ctx->ssa;
+	int free_var = -1;
+	zend_uchar free_var_type;
+
+	if (opline->opcode == ZEND_NOP) {
+		return 0;
+	}
+
+	/* We mark FREEs as dead, but they're only really dead if the destroyed var is dead */
+	if (opline->opcode == ZEND_FREE && !is_var_dead(ctx, ssa_op->op1_use)) {
+		return 0;
+	}
+
+	if ((opline->op1_type & (IS_VAR|IS_TMP_VAR)) && !is_var_dead(ctx, ssa_op->op1_use)) {
+		if (!try_remove_var_def(ctx, ssa_op->op1_use, ssa_op->op1_use_chain, opline)) {
+			free_var = ssa_op->op1_use;
+			free_var_type = opline->op1_type;
+		}
+	}
+	if ((opline->op2_type & (IS_VAR|IS_TMP_VAR)) && !is_var_dead(ctx, ssa_op->op2_use)) {
+		if (!try_remove_var_def(ctx, ssa_op->op2_use, ssa_op->op2_use_chain, opline)) {
+			if (free_var >= 0) {
+				// TODO: We can't free two vars. Keep instruction alive.
+				zend_bitset_excl(ctx->instr_dead, opline - ctx->op_array->opcodes);
+				return 0;
+			}
+			free_var = ssa_op->op2_use;
+			free_var_type = opline->op2_type;
+		}
+	}
+
+	zend_ssa_rename_defs_of_instr(ctx->ssa, ssa_op);
+	zend_ssa_remove_instr(ctx->ssa, opline, ssa_op);
+
+	if (free_var >= 0) {
+		opline->opcode = ZEND_FREE;
+		opline->op1.var = (uintptr_t) ZEND_CALL_VAR_NUM(NULL, ssa->vars[free_var].var);
+		opline->op1_type = free_var_type;
+
+		ssa_op->op1_use = free_var;
+		ssa_op->op1_use_chain = ssa->vars[free_var].use_chain;
+		ssa->vars[free_var].use_chain = ssa_op - ssa->ops;
+		return 0;
+	}
+	return 1;
+}
+
+// TODO Move this somewhere else (CFG simplification?)
+static int simplify_jumps(zend_ssa *ssa, zend_op_array *op_array) {
+	int removed_ops = 0;
+	zend_basic_block *block;
+	FOREACH_BLOCK(block) {
+		int block_num = block - ssa->cfg.blocks;
+		zend_op *opline = &op_array->opcodes[block->start + block->len - 1];
+		zend_ssa_op *ssa_op = &ssa->ops[block->start + block->len - 1];
+		zval *op1;
+
+		if (block->len == 0) {
+			continue;
+		}
+
+		/* Convert jump-and-set into jump if result is not used */
+		switch (opline->opcode) {
+			case ZEND_JMPZ_EX:
+				ZEND_ASSERT(ssa_op->result_def >= 0);
+				if (ssa->vars[ssa_op->result_def].use_chain < 0
+						&& ssa->vars[ssa_op->result_def].phi_use_chain == NULL) {
+					opline->opcode = ZEND_JMPZ;
+					opline->result_type = IS_UNUSED;
+					zend_ssa_remove_result_def(ssa, ssa_op);
+				}
+				break;
+			case ZEND_JMPNZ_EX:
+			case ZEND_JMP_SET:
+				ZEND_ASSERT(ssa_op->result_def >= 0);
+				if (ssa->vars[ssa_op->result_def].use_chain < 0
+						&& ssa->vars[ssa_op->result_def].phi_use_chain == NULL) {
+					opline->opcode = ZEND_JMPNZ;
+					opline->result_type = IS_UNUSED;
+					zend_ssa_remove_result_def(ssa, ssa_op);
+				}
+				break;
+		}
+
+		/* Convert jump-and-set to QM_ASSIGN/BOOL if the "else" branch is not taken. */
+		switch (opline->opcode) {
+			case ZEND_JMPZ_EX:
+			case ZEND_JMPNZ_EX:
+				if (block->successors[1] < 0 && block->successors[0] != block_num + 1) {
```

Comment:
Not totally sure, but I think this should check for `block->successors_count == 1` instead of `block->successors[1] < 0` nowadays.